src/index.ts

```diff
@@ -0,0 +1,221 @@
+  follow?: boolean;
+}
+
+interface IDockerComposeBuildOptions extends IDockerComposeOptions {
+  parallel?: boolean;
+}
+
+interface IDockerComposePushOptions extends IDockerComposeOptions {
+  ignorePushFailures?: boolean;
+}
+
+interface IDockerComposeResult {
+  exitCode: number | null;
+  out: string;
+  err: string;
+}
+
+/**
+ * Converts supplied yml files to cli arguments
+ * https://docs.docker.com/compose/reference/overview/#use--f-to-specify-name-and-path-of-one-or-more-compose-files
+ */
+const configToArgs = (config): string[] => {
+  if (typeof config === 'undefined') {
+    return [];
+  } else if (typeof config === 'string') {
+    return [ '-f', config ];
+  } else if (config instanceof Array) {
+    return config.reduce((args, item): string[] => args.concat([ '-f', item ]), []);
+  }
+  throw new Error(`Invalid argument supplied: ${config}`);
+};
+
+/**
+ * Converts docker-compose commandline options to cli arguments
+ */
+const composeOptionsToArgs = (composeOptions): string[] => {
+  let composeArgs: string[] = [];
+
+  composeOptions.forEach((option: string[] | string): void => {
+    if (option instanceof Array) {
+      composeArgs = composeArgs.concat(option);
+    }
+    if (typeof option === 'string') {
+      composeArgs = composeArgs.concat([ option ]);
+    }
+  });
+
+  return composeArgs;
+};
+
+/**
+ * Executes docker-compose command with common options
+ */
+const execCompose = (command, args, options: IDockerComposeOptions = {}): Promise<IDockerComposeResult> => new Promise((resolve, reject): void => {
+  const composeOptions = options.composeOptions || [];
+  const commandOptions = options.commandOptions || [];
+  let composeArgs = composeOptionsToArgs(composeOptions);
+
+  composeArgs = composeArgs.concat(configToArgs(options.config).concat([ command ].concat(composeOptionsToArgs(commandOptions), args)));
+
+  const cwd = options.cwd;
+  const env = options.env || undefined;
+
+  const childProc = childProcess.spawn('docker-compose', composeArgs, { cwd, env });
+
+  childProc.on('error', (err): void => {
+    reject(err);
+  });
+
+  const result: IDockerComposeResult = {
+    exitCode: null,
+    err: '',
+    out: ''
+  };
+
+  childProc.stdout.on('data', (chunk): void => {
+    result.out += chunk.toString();
+  });
+
+  childProc.stderr.on('data', (chunk): void => {
+    result.err += chunk.toString();
+  });
+
+  childProc.on('exit', (exitCode): void => {
+    result.exitCode = exitCode;
+    if (exitCode === 0) {
+      resolve(result);
+    } else {
+      reject(result);
+    }
+  });
+
+  if (options.log) {
+    childProc.stdout.pipe(process.stdout);
+    childProc.stderr.pipe(process.stderr);
+  }
+});
+
+export const upAll = function (options?: IDockerComposeOptions): Promise<IDockerComposeResult> {
+  return execCompose('up', [ '-d' ], options);
+};
+
+export const upMany = function (services, options?: IDockerComposeOptions): Promise<IDockerComposeResult> {
```

Comment:
using `services: string[]` will make `index.d.ts` more similar to the old one

Reply:
Whoops. Yeah, it should have been defined. I missed that one when defining types.

There are also a couple other internal functions that would benefit from strict typing, but I've left them untouched for the moment - I wasn't 100% sure if I'd identify types correctly (which is kind of the point of TypeScript and should cause an error, but I wasn't sure)..

Reply:
Updated accordingly.